### PR TITLE
refactor(core): Centralize shell logic into ShellExecutionService

### PR DESCRIPTION
**Pull request by abhipatel12**
## TLDR

This is a non-user-facing refactor that introduces a new, centralized `ShellExecutionService` in `@google/gemini-cli-core`. **There are no functional or user-facing changes in this pull request.**

Previously, complex shell execution logic (process spawning, cross-platform termination, output streaming) was duplicated in two places:

1.  The interactive `!` shell processor (`packages/cli/src/ui/hooks/shellCommandProcessor.ts`)
2.  The `run_shell_command` tool (`packages/core/src/tools/shell.ts`)

This PR extracts that duplicated logic into the new service and refactors both `shellCommandProcessor` and `ShellTool` to be simple, clean clients of this service.

**Reviewers should focus on:**

*   The clean API of the new `ShellExecutionService`.
*   The simplification of `shellCommandProcessor.ts` and `shell.ts`.
*   The comprehensive new unit tests that now cover this critical logic in isolation.

## Dive Deeper

This refactor is a strategic investment in the stability, security, and maintainability of our shell interaction layer. The primary architectural benefits are:

*   **DRY (Don't Repeat Yourself):** We have eliminated significant code duplication. A bug fix or improvement to our shell handling (e.g., improving process termination on Windows) now only needs to be made in one place.

*   **Single Source of Truth:** The `ShellExecutionService` becomes the canonical implementation for how Gemini CLI interacts with the system shell. This guarantees consistent behavior, whether a command is invoked interactively by the user (`!ls`) or programmatically by the model (`run_shell_command`).

*   **Improved Testability:** The most complex, platform-dependent logic is now in a pure, dependency-free class. This allowed us to write a robust and isolated test suite (`shellExecutionService.test.ts`) that covers edge cases like process cancellation, binary output, and platform-specific termination paths—scenarios that were previously untestable in a clean, hermetic way.

*   **Enabling Future Features:** This refactor is a direct prerequisite for upcoming features, such as adding `!` shell execution capabilities to custom commands. The new `ShellProcessor` for custom commands will become the third client of this service, allowing for rapid and secure implementation.

The new service also introduces a more robust, event-driven API (`onOutputEvent`) for consumers to react to different types of output (text data, binary detection, binary progress), which is a technical improvement over the previous implementation.

## Reviewer Test Plan

The primary validation for this refactor is that all existing functionality remains unchanged.

#### 1. Automated Tests (Primary Validation)

The most important step is to run the full test suite. All new and existing tests must pass.

```bash
npm test
```

Pay special attention to the new test files, which provide confidence in this refactor:
*   `packages/core/src/services/shellExecutionService.test.ts`
*   `packages/core/src/tools/shell.test.ts` (the `execute` block is all new)
*   `packages/cli/src/ui/hooks/shellCommandProcessor.test.ts` (completely rewritten)

#### 2. Manual Sanity Checks

After pulling the branch, run the `gemini` command in interactive mode and manually verify the following common shell scenarios. The behavior should be **identical** to the `main` branch.

1.  **Simple Success:** Run `!ls -l` (or `!dir` on Windows).
    *   **Expected:** It should list files successfully.

2.  **Stderr Output:** Run `!a-bad-command`.
    *   **Expected:** It should display an error from the shell (e.g., "command not found").

3.  **Command Cancellation:** Run `!sleep 10` and immediately press `Ctrl+C`.
    *   **Expected:** The command should be canceled, and the UI should reflect this.

4.  **Tool-based Execution:** Start a conversation and ask the model to perform a shell action.
    *   **Prompt:** `"Please list the files in the current directory for me."`
    *   **Expected:** The model should propose using the `run_shell_command` tool, which you can then approve. The result should be displayed correctly.

5.  **Directory Change Warning (Non-Windows):** Run `!cd .. && ls`.
    *   **Expected:** The command should execute, but you should see a warning in the final output: `WARNING: shell mode is stateless; the directory change to '...' will not persist.`

## Testing Matrix

|          | 🍏  | 🪟  | 🐧  |
| -------- | --- | --- | --- |
| npm run  | ✅  | ❓  | ❓  |
| npx      | ❓  | ❓  | ❓  |
| Docker   | ❓  | ❓  | ❓  |
| Podman   | ❓  | -   | -   |
| Seatbelt | ✅  | -   | -   |

## Linked issues / bugs

This PR makes progress on #3789

### CODE REVIEW

File: packages/core/src/tools/shell.ts [hunk 1]
```diff
@@ -313,122 +314,76 @@
...
-    }
+        if (shouldUpdate) {
+          updateOutput(currentDisplayOutput);
+          lastUpdateTime = Date.now();
+        }
+      },
+      abortSignal,
+    );
+
+    const result = await resultPromise;
```
[the remark this answers is not in the document]
Comment (jacob314) [reply packages/core/src/tools/shell.ts]: agree this would be good. feel free to add in a follow up pull request as it is not a new issue.

File: packages/core/src/services/shellExecutionService.ts [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+  static execute(
+    commandToExecute: string,
+    cwd: string,
+    onOutputEvent: (event: ShellOutputEvent) => void,
+    abortSignal: AbortSignal,
+  ): ShellExecutionHandle {
+    const isWindows = os.platform() === 'win32';
+    const shell = isWindows ? 'cmd.exe' : 'bash';
+    const shellArgs = isWindows
+      ? ['/c', commandToExecute]
```
Comment (jacob314) [inline packages/core/src/services/shellExecutionService.ts]: tiny nit: move the isWindows check inside the array
`[isWindows ? '/c' : '0', commandToExecute]`

File: packages/core/src/utils/textUtils.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+ */
+
+/**
+ * Checks if a Buffer is likely binary by testing for the presence of a NULL byte.
+ * The presence of a NULL byte is a strong indicator that the data is not plain text.
+ * @param data The Buffer to check.
+ * @param sampleSize The number of bytes from the start of the buffer to test.
+ * @returns True if a NULL byte is found, false otherwise.
+ */
+export function isBinary(
```
Comment (jacob314) [inline packages/core/src/utils/textUtils.ts:14]: there is already an isBinary util in textUtils.ts in the cli package.
I would suggest merging that method with this one and removing it from the cli textUtils package. for discussion, it might make sense to create  a separate utils package used by both core and cli rather than defining utils in core that are then referenced by both core and cli.

File: packages/cli/src/ui/hooks/shellCommandProcessor.ts [hunk 4]
```diff
@@ -285,26 +122,82 @@ export const useShellCommandProcessor = (
...
                 ],
               });
               lastUpdateTime = Date.now();
             }
           },
-          onDebugMessage,
-        )
-          .then((result) => {
+          abortSignal,
+        );
```
[the remark this answers is not in the document]
Comment (jacob314) [reply packages/cli/src/ui/hooks/shellCommandProcessor.ts]: not a new issue but agree worth fixing even if the error is unlikely

File: packages/core/src/services/shellExecutionService.ts [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+      const abortHandler = async () => {
+        if (child.pid && !exited) {
+          if (isWindows) {
+            spawn('taskkill', ['/pid', child.pid.toString(), '/f', '/t']);
+          } else {
+            try {
+              // Kill the entire process group (negative PID).
+              // SIGTERM first, then SIGKILL if it doesn't die.
+              process.kill(-child.pid, 'SIGTERM');
+              await new Promise((res) => setTimeout(res, 200));
```
Comment (jacob314) [inline packages/core/src/services/shellExecutionService.ts]: 200 is a bit of a magic number. make it a const while you are cleaning this up.

**Review by jacob314 — APPROVED** [review]
lgtm
before landing please normalize the isBinary methods likely picking the one from the cli package that looks a bit more advanced.

File: packages/cli/src/ui/hooks/shellCommandProcessor.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (abhipatel12) [reply packages/cli/src/ui/hooks/shellCommandProcessor.ts]: Done

File: packages/core/src/services/shellExecutionService.ts (hunk 2 above)
Comment (abhipatel12) [reply packages/core/src/services/shellExecutionService.ts]: Done

File: packages/core/src/services/shellExecutionService.ts (hunk 5 above)
Comment (abhipatel12) [reply packages/core/src/services/shellExecutionService.ts]: Done

File: packages/core/src/tools/shell.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (abhipatel12) [reply packages/core/src/tools/shell.ts]: Done

File: packages/core/src/utils/textUtils.ts (hunk 3 above)
Comment (abhipatel12) [reply packages/core/src/utils/textUtils.ts:14]: Done

**Comment by abhipatel12** [comment]
Thanks for the review! Made all the changes
